// labels/agenda-settings/agendaEdition.js
"use strict";

module.exports = {
  description: {
    fr: 'Description',
    en: 'Description'
  },
  name: {
    fr: 'Nom',
    en: 'Name'
  },
  namePlaceholder: {
    fr: 'Mon organisation, mon festival, ma ville, autre',
    en: 'My organization, my festival, my city, other'
  },
  website: {
    fr: 'Site web',
    en: 'Website'
  },
  websitePlaceholder: {
    fr: 'http://monsite.com',
    en: 'http://mywebsite.com'
  },
  personalizedSlug: {
    fr: 'Adresse web personnalisée',
    en: 'Personalized web address'
  },
  contribution: {
    fr: 'Contribution',
    en: 'Contribution'
  },
  contribType: {
    fr: 'Qui peut ajouter des événements à cet agenda ?',
    en: 'Who can add events to this calendar ?'
  },
  contribTypeChoosen: {
    fr: 'Seulement des personnes choisies et moi',
    en: 'Only selected people and me'
  },
  contribTypeAll: {
    fr: 'Tout le monde',
    en: 'Everybody'
  },
  contribDefaultState: {
    fr: 'Statut par défaut des événements',
    en: 'Default status of events'
  },
  contribDefaultStatePublished: {
    fr: 'Publiés',
    en: 'Published'
  },
  contribDefaultStatePublishedText: {
    fr: 'vous pouvez modifier les événements ensuite',
    en: 'you can edit the events later'
  },
  contribDefaultStateUnpublished: {
    fr: 'Non-publiés',
    en: 'Unpublished'
  },
  contribDefaultStateUnpublishedText: {
    fr: 'les événements ne deviennent publics qu\'après votre validation',
    en: 'events become public only after your validation'
  },
  'required': {
    fr: 'Requis',
    en: 'Required'
  },
  'string.tooshort': {
    fr: 'La chaîne de caractères est trop courte',
    en: 'String is too short'
  },
  'string.toolong': {
    fr: 'La chaîne de caractères est trop longue',
    en: 'String is too long'
  },
  'link.invalid': {
    fr: 'Lien invalide',
    en: 'Invalid link'
  },
  duplicate: {
    fr: 'Cette valeur est déjà prise par un autre agenda',
    en: 'Duplicate value found'
  },
  saveModifications: {
    fr: 'Sauvegarder les modifications',
    en: 'Save changes'
  },
  saving: {
    fr: 'Sauvegarde en cours',
    en: 'Saving...'
  },
  saved: {
    fr: 'Modifications sauvegardées',
    en: 'Changes saved'
  },
  agendaProfile: {
    fr: 'Profil de l\'agenda',
    en: 'Agenda profile'
  },
  consigne: {
    fr: 'Consigne',
    en: 'Guideline'
  },
  consigneSubLabel: {
    fr: 'La consigne s\'affiche au début du formulaire de saisie d\'événement',
    en: 'The guideline appears at the beginning of the event entry form'
  },
  yes: {
    fr: 'Oui',
    en: 'Yes'
  },
  no: {
    fr: 'Non',
    en: 'No'
  },
  contribUseFields: {
    fr: 'Inviter chaque membre à se présenter (organisation, téléphone, nom, prénom, fonction, email)',
    en: 'Invite members to present themselves (organization, phone, name, title, email)'
  },
  removeAgenda: {
    fr: 'Supprimer l\'agenda',
    en: 'Delete agenda'
  },
  removeAgendaWarning: {
    fr: 'Êtes-vous sûr de vouloir supprimer cet agenda ?',
    en: 'Are you sure you want to remove this agenda ?'
  },
  close: {
    fr: 'Fermer',
    en: 'Close'
  },
  remove: {
    fr: 'Supprimer',
    en: 'Remove'
  },
  agendaRemoved: {
    fr: 'Votre agenda a bien été supprimé',
    en: 'Your agenda has been deleted'
  }
};
