// labels/event/show.js
module.exports = {
  addToAgenda: {
    en: 'Add to my agenda',
    fr: 'Ajouter à mon agenda',
  },
  export: {
    en: 'Export',
    fr: 'Exporter'
  },
  share: {
    en: 'Share',
    fr: 'Partager',
  },
  back: {
    en: 'Back',
    fr: 'Retour'
  },
  ticketingLink: {
    en: 'Ticketing link',
    fr: 'Lien d\'inscription',
  },
  startToFinish: {
    en: '%start% to %finish%',
    fr: 'du %start% au %finish%',
  },
  onAt: {
    en: 'on %start% @ %time%',
    fr: 'le %start% @ %time%',
  },
  draft: {
    en: 'Draft',
    fr: 'Brouillon',
  },
  published: {
    en: 'Published',
    fr: 'Publié',
  },
  private: {
    en: 'Private',
    fr: 'Privé',
  },
  edit: {
    en: 'Edit',
    fr: 'Modifier',
  },
  delete: {
    en: 'Delete',
    fr: 'Supprimer',
  },
  about: {
    en: 'About',
    fr: 'A propos',
  },
  help: {
    en: 'Help',
    fr: 'Aide',
  },
  termsOfUse: {
    en: 'Terms of use',
    fr: 'Conditions d\'utilisation',
  },
  descriptionAvailableIn: {
    en: 'Description available in',
    fr: 'Description disponible en',
  },
  dateDetails: {
    en: 'Date details',
    fr: 'Détail des dates',
  },
  createdBy: {
    en: 'Created by %owner%',
    fr: 'Créé par %owner%',
  },
  pastEvent: {
    en: 'Past event',
    fr: 'Événement passé',
  },
  seeAllEvents: {
    en: 'see all the events',
    fr: 'voir tous les événements',
  },
  areYouSureToDelete: {
    en: 'Are you sure you want to delete this event?',
    fr: 'Etes-vous sûr de vouloir supprimer cet événement?',
  },
  confirm: {
    en: 'Confirm',
    fr: 'Confirmer',
  },
  cancel: {
    en: 'Cancel',
    fr: 'Annuler',
  },
  eventState: {
    en: 'Event state',
    fr: 'Etat de l\'événement',
  },
  toComplete: {
    en: 'To complete',
    fr: 'A compléter',
  },
  readyForPublishing: {
    en: 'Ready for publishing',
    fr: 'Prêt à publier',
  },
  unpublished: {
    en: 'Unpublished',
    fr: 'Non publié',
  },
  backToDashboard: {
    en: 'Back to the dashboard',
    fr: 'Retour à la gestion des événements',
  },
  changeFeaturedState: {
    en: 'Click here to change the featured state',
    fr: 'Cliquez ici pour changer l\'état \'En une\'',
  },
  addToFeatured: {
    en: 'Add to featured',
    fr: 'Mettre en une',
  },
  removeFromFeatured: {
    en: 'Remove from featured',
    fr: 'Retirer de la une',
  },
  featured: {
    en: 'Featured',
    fr: 'Mis en une',
  },
  fromTo: {
    en: 'from %from% to %to%',
    fr: 'de %from% à %to%',
  },
  duplicate: {
    en: 'Duplicate',
    fr: 'Dupliquer',
  },
  startingAt: {
    en: 'Starting at %min% years old',
    fr: 'A partir de %min% ans',
  },
  minToMaxYearsOld: {
    en: '%min% to %max% years old',
    fr: 'De %min% à %max% ans',
  },
  accessibleToHearing: {
    en: 'Accessible to the hearing impaired',
    fr: 'Accessible aux handicapés auditifs',
  },
  accessibleToMentally: {
    en: 'Accessible to the mentally impaired',
    fr: 'Accessible aux handicapés psychiques',
  },
  accessibleToMotor: {
    en: 'Accessible to the motor impaired',
    fr: 'Accessible aux handicapés moteurs',
  },
  accessibleToVisually: {
    en: 'Accessible to the visually impaired',
    fr: 'Accessible aux handicapés visuels',
  },
  signLanguageAvailable: {
    en: 'Sign language available',
    fr: 'Accessible en langue des signes',
  },
  accessibilityImpairments: {
    en: 'Accessibility: %impairments%',
    fr: 'Accessibilité: %impairments%',
  },
  transferOwnership: {
    en: 'Transfer ownership',
    fr: 'Transférer la propriété',
  },
  access: {
    en: 'Access',
    fr: 'Accès',
  },
  aboutTheVenue: {
    en: 'About the venue',
    fr: 'A propos du lieu',
  },
  tags: {
    en: 'Tags',
    fr: 'Tags',
  },
  category: {
    en: 'Category',
    fr: 'Catégorie',
  },
  contact: {
    en: 'Contact',
    fr: 'Contacter',
  },
  requestChangeLocation: {
    en: 'Request a change in the location information',
    fr: 'Faire une demande de modification du lieu',
  },
  locationChangeRequest: {
    en: 'Location information change request: %name%',
    fr: 'Demande de modification du lieu: %name%',
  },
  couldYouBringFollowingChanges: {
    en: 'Hi, could you bring the following changes to the location named \'%name\'? reference: %uid% - ( add details of the change here )',
    fr: 'Bonjour, pourriez-vous apporter les modifications suivantes au lieu ayant pour nom \'%name%\' ? reference: %uid% - ( détaillez les modifications ici )',
  },
  moreLinks: {
    en: 'More links',
    fr: 'Liens additionnels',
  },
  state: {
    en: 'State',
    fr: 'Statut',
  },
  editLocation: {
    en: 'Edit location',
    fr: 'Editer le lieu',
  },
  actions: {
    en: 'Actions',
    fr: 'Actions'
  },
  activity: {
    en: 'Activity',
    fr: 'Activité'
  }
};
