// labels/call-to-action/index.js
"use strict";

module.exports = {
  activationRequest: {
    fr: 'Demande d\'activation',
    en: 'Activation request'
  },
  message: {
    fr: 'Message',
    en: 'Message'
  },
  send: {
    fr: 'Envoyer',
    en: 'Send'
  },
  confirmationMessage: {
    fr: 'Message de confirmation',
    en: 'Confirmation message'
  },
  requestSuccess: {
    fr: 'Votre demande a bien été envoyé.',
    en: 'Your request has been sent.'
  },
  requestFail: {
    fr: 'Une erreur a eu lieu lors de l\'envoi de votre demande, veuillez réessayer ultérieurement.',
    en: 'An error occurred while sending your request, please try again later.'
  },

  requestTitleInvitationMessage: {
    fr: 'Demande d\'activation des messages personnalisés',
    en: 'Request to activate personalized messages'
  },
  requestMessageInvitationMessage: {
    fr: 'Cette fonctionnalité vous permet d\'inviter un/des utilisateur(s) avec un message personnalisé,' +
    ' ainsi qu\'écrire un message à tous les membres.',
    en: 'This feature allows you to invite a user(s) with a custom message, and writing a message to all members.'
  },

  requestTitlePrivateAgenda: {
    fr: 'Demande d\'activation d\'un agenda privé',
    en: 'Request to activate a private agenda'
  },
  requestMessagePrivateAgenda: {
    fr: 'Un OpenAgenda privé n\'est accessible que par ses membres. \n\nIl n\'est pas indexé : ni par l\'API OpenAgenda, ni par le moteur de recherche OpenAgenda, ni par Google (et autres moteurs de recherche).',
    en: 'A private OpenAgenda is accessible only by its members. \n\nIt is not indexed: neither by the OpenAgenda API, nor by the OpenAgenda search engine, nor by Google (and other search engines).'
  }
}
