// labels/members/index.js
module.exports = {
  members: {
    fr: 'Membres',
    en: 'Members'
  },
  searchMember: {
    fr: 'Rechercher un membre',
    en: 'Search a member'
  },
  contributor: {
    fr: 'Contributeur',
    en: 'Contributor'
  },
  administrator: {
    fr: 'Administrateur',
    en: 'Administrator'
  },
  moderator: {
    fr: 'Modérateur',
    en: 'Moderator'
  },
  reader: {
    fr: 'Lecteur',
    en: 'Reader'
  },
  contributors: {
    fr: 'Contributeurs',
    en: 'Contributors'
  },
  administrators: {
    fr: 'Administrateurs',
    en: 'Administrators'
  },
  moderators: {
    fr: 'Modérateurs',
    en: 'Moderators'
  },
  readers: {
    fr: 'Lecteurs',
    en: 'Readers'
  },
  events: {
    fr: 'événements',
    en: 'events'
  },
  event: {
    fr: 'événement',
    en: 'event'
  },
  result: {
    fr: 'Résultat',
    en: 'Result'
  },
  noResult: {
    fr: 'Aucun résultat trouvé',
    en: 'No result found'
  },
  actions: {
    fr: 'Actions',
    en: 'Actions'
  },
  removeMember: {
    fr: 'Retirer des membres',
    en: 'Remove member'
  },
  removeConfirmMessage: {
    fr: 'Êtes-vous sûr de vouloir supprimer cet utilisateur des membres ?',
    en: 'Are you sure you want to delete this user from members ?'
  },
  removeModalError: {
    fr: 'Vous ne pouvez pas supprimer ce membre.',
    en: 'You canno\'t remove this member.'
  },
  editProfile: {
    fr: 'Modifier le profil',
    en: 'Edit profile'
  },
  inviteMembers: {
    fr: 'Inviter des membres',
    en: 'Invite members'
  },
  inviteMembersPlaceholder: {
    fr: 'Collez ici les adresses email des nouveaux membres, séparées par des virgules',
    en: 'Paste here the email addresses of new members, separated by commas'
  },
  emails: {
    fr: 'Emails',
    en: 'Emails'
  },
  role: {
    fr: 'Rôle',
    en: 'Role'
  },
  selectRole: {
    fr: 'Sélectionnez un rôle',
    en: 'Select a role'
  },
  name: {
    fr: 'Nom',
    en: 'Name'
  },
  email: {
    fr: 'Email',
    en: 'Email'
  },
  phone: {
    fr: 'Téléphone',
    en: 'Phone'
  },
  position: {
    fr: 'Poste',
    en: 'Position'
  },
  organization: {
    fr: 'Organisation',
    en: 'Organization'
  },
  total: {
    fr: 'Total',
    en: 'Total'
  },
  required: {
    fr: 'Requis',
    en: 'Required'
  },
  'string.tooshort': {
    fr: 'La chaîne de caractères est trop courte',
    en: 'String is too short'
  },
  'phone.invalid': {
    fr: 'Numéro de téléphone invalide',
    en: 'Phone number is invalid'
  },
  'email.invalid': {
    fr: 'L\'adresse email est invalide',
    en: 'Email address is invalid'
  },
  'emails.invalid': {
    fr: 'Au moins une des adresses email est invalide',
    en: 'At least one of the email addresses is invalid'
  },
  'credential.downgrade': {
    fr: 'Vous ne pouvez pas baisser le niveau d\'un membre',
    en: 'You can not downgrade credential of a member'
  },
  exportToXlsx: {
    fr: 'Exporter en xlsx',
    en: 'Export to xlsx'
  },
  exportToCsv: {
    fr: 'Exporter en csv',
    en: 'Export to csv'
  },
  writeToHim: {
    fr: 'Lui écrire',
    en: 'Write to him'
  },
  emailsCouldNotBeInvited: {
    fr: 'Les emails suivants n\'ont pas pu être invités:',
    en: 'The following emails could not be invited:'
  },
  invitationProblem: {
    fr: 'Il y a eu un problème pendant l\'invitation des membres, veuillez réessayer ultérieurement.',
    en: 'There was a problem while inviting your members, please try again later.'
  },
  membersInvited: {
    fr: 'Les membres ont été invités avec succès.',
    en: 'Members were invited successfully.'
  },
  noName: {
    fr: 'Sans nom',
    en: 'Nameless'
  },
  deletedUser: {
    fr: 'Utilisateur supprimé',
    en: 'Deleted user'
  },
  invited: {
    fr: 'Invité',
    en: 'Invited'
  },
  resendInvitation: {
    fr: 'Relancer l\'invitation',
    en: 'Resend invitation'
  },
  invitationResended: {
    fr: 'L\'invitation a été renvoyée.',
    en: 'The invitation was resent.'
  },
  invitationNotResended: {
    fr: 'L\'invitation n\'a pas pu être renvoyée.',
    en: 'The invitation could not be resent.'
  },
  message: {
    fr: 'Message',
    en: 'Message'
  },
  replyTo: {
    fr: 'Adresse email de réponse',
    en: 'Response email address'
  },
  active: {
    fr: 'Actif',
    en: 'Active'
  },
  inactive: {
    fr: 'Inactif',
    en: 'Inactive'
  },
  noContrib: {
    fr: 'n\'a pas contribué',
    en: 'did not contribute'
  },
  deleted: {
    fr: 'Supprimé',
    en: 'Deleted'
  },
  sendMessage: {
    fr: 'Envoyer le message',
    en: 'Send message'
  },
  sendMessageToAll: {
    fr: 'Envoyer un message à tous les membres',
    en: 'Send a message to all members'
  },
  sendMessageToInactives: {
    fr: 'Envoyer un message aux membres n\'ayant pas contribué',
    en: 'Send a message to members who have not contributed'
  },
  messageSent: {
    fr: 'Le message a été bien envoyé.',
    en: 'The message was sent successfully.'
  },
  close: {
    fr: 'Fermer',
    en: 'Close'
  },
  moreinfoActive: {
    fr: 'Utilisateur membre de l\'agenda ayant contribué des événements',
    en: 'Member that has contributed events'
  },
  moreinfoInactive: {
    fr: 'Utilisateur membre de l\'agenda n\'ayant pas encore contribué d\'événements',
    en: 'Member that hasn\'t contributed any event yet'
  },
  moreinfoInvited: {
    fr: 'Utilisateur ayant été invité à devenir membre',
    en: 'User that has been invited to become member'
  },
  moreinfoDeleted: {
    fr: 'Membre ayant supprimé son compte utilisateur',
    en: 'Member that has deleted his/her user account'
  }
};
