// cibulModel/lib/events.js
module.exports = function(m, f) {

  model = m;

  fileHandler = f;

  return {
    get: get,
    create: create,
    list: list,
    clear: clear,
    createSingleDateEvent: createSingleDateEvent,
    getUpcoming: getUpcoming,
    basicCheckExists: basicCheckExists,
    getByLocationId: getByLocationId
  };
  
};

var model, 

getQuery = [
  'select e.created_at as created_at, ',
  'e.updated_at as updated_at,',
  'e.slug as slug,',
  'e.uid as uid,',
  'e.owner_id as owner_id,',
  'e.image as image,',
  'e.id as id,',
  't.lang as lang,',
  't.title as title,',
  't.description as description,',
  'l.city as city, l.department as department, l.region as region, l.city_district as district , l.postal_code as postcode,',
  't.tags as tags,',
  't.free_text as free_text,',
  'o.date as date,',
  'o.time_start as time_start,',
  'o.time_end as time_end,',
  'l.placename as placename,',
  'l.latitude as latitude,',
  'l.longitude as longitude,',
  'l.address as address,',
  'l.id as location_id,',
  'l.uid as location_uid,',
  'l.main as location_verified,',
  'el.ticket_link as ticket_link,',
  'elt.pricing_info as pricing_info,',
  'elt.lang as ellang',
  'from event as e left join event_translation as t on e.id=t.id',
  'left join event_location as el on e.id=el.event_id',
  'left join occurrence as o on el.event_id=o.event_id and el.location_id=o.location_id',
  'left join location as l on el.location_id=l.id',
  'left join event_location_translation as elt on el.id=elt.id',
  'where e.is_new=false and e.is_published=true'
].join(' '),

lib = require('./lib'),

async = require('async'),

image = require('./image'),

thumbnailPrefix = 'evtb',

s3Path = '//cibul.s3.amazonaws.com/',

fileHandler,

/**
 * clear the events set
 */

clear = function(cb) {

  model.clear([
    'conversation_event_editor_request_info',
    'conversation_event_claim_info',
    'conversation_signal_event_info',
    'event', 
    'event_translation', 
    'event_location',
    'event_location_translation', 
    'occurrence',
    'event_activity',                          // deprecated
    'event_comment',                           // deprecated
  ], cb);

},

validateEventEntry = function(values) {

  if (!values.owner_id) throw 'no owner id is specified';

},

addEventEntry = function(title, values, cb) {

  _createSlugAndUid(title, function(err, result) {

    validateEventEntry(values);

    var con = model.getConnection(),

    entry = {
      slug: result.slug,
      uid: result.uid,
      created_at: new Date(),
      updated_at: new Date(),
      is_published: true,
      is_new: false,
      owner_id: values.owner_id,
    };

    if (values.image) entry.image = values.image;

    con.query('insert into event set ?', entry, function(err, result) {

      if (err) throw err;

      entry.id = result.insertId;

      cb(entry);

    });

  });

},

validateEventTranslationEntry = function(values) {

  if (!values.title) throw 'title is not defined';
  if (!values.lang) throw 'lang is not defined';

},

addEventTranslationEntry = function(values, cb) {

  validateEventTranslationEntry(values);

  if (!values.id) throw 'id is not defined';

  var con = model.getConnection(),

  entry = {
    title: values.title,
    lang: values.lang,
    id: values.id
  };

  if (values.description) entry.description = values.description;
  if (values.free_text) entry.free_text = values.free_text;
  if (values.tags) entry.tags = values.tags;

  con.query('insert into event_translation set ?', entry, function(err, result) {

    if (err) throw err;

    cb(entry);

  });

},

validateOccurrenceEntry = function(values) {

  if (!values.start) throw 'start datetime is not defined';
  if (!values.start instanceof Date) throw 'start should be a Date';
  if (values.end && !values.end instanceof Date) throw 'end should be a Date';
  if (values.end && ((values.end-values.start)/1000/24/60/60 >= 1)) throw 'end cannot be more than 24 hours after start';
  if (!values.location_id) throw 'location_id is not defined';

},

addOccurrenceEntry = function(values, cb) {

  validateOccurrenceEntry(values);

  if (!values.event_id) throw 'event_id is not defined';
  
  var con = model.getConnection(),

  entry = {
    date: values.start.getFullYear() + '/' + lib.addZero(values.start.getMonth() + 1) + '/' + lib.addZero(values.start.getDate()),
    time_start: lib.addZero(values.start.getHours()) + ':' + lib.addZero(values.start.getMinutes()) + ':00',
    time_end: values.end?addZero(values.end.getHours()) + ':' + lib.addZero(values.end.getMinutes()) + ':00':'00:00:00',
    event_id: values.event_id,
    location_id: values.location_id,
    created_at: new Date(),
    updated_at: new Date()
  };

  con.query('insert into occurrence set ?', entry, function(err, result) {

    if (err) throw err;

    entry.id = result.insertId;

    cb(entry);

  });

},

addEventLocationEntry = function(values, cb) {

  if (!values.event_id) throw 'EventLocation: event_id is needed';
  if (!values.location_id) throw 'EventLocation: location_id is needed';

  var con = model.getConnection(),

  entry = { 
    event_id: values.event_id, 
    location_id: values.location_id,
    created_at: new Date(),
    updated_at: new Date()
  };

  con.query('insert into event_location set ?', entry, function(err, result) {

    if (err) throw err;

    entry.id = result.insertId;

    cb(entry);

  });

},

addEventLocationTranslationEntry = function(values, cb) {

  if (!values.event_location_id) throw 'EventLocationTranslation: event_location_id is needed';

  var con = model.getConnection(),

  entry = { 
    id: values.event_location_id,
    pricing_info: values.pricing?values.pricing:''
  };

  con.query('insert into event_location_translation set ?', entry, function(err, result) {

    if (err) throw err;

    entry.id = result.insertId;

    cb(entry);

  });

},

validateSingleDateEvent = function(event) {

  var compulsoryFields = ['title', 'description', 'start', 'locationId', 'ownerId'];

  for (var i=0; i<compulsoryFields.length;i++)
    if (event[compulsoryFields[i]] == undefined) throw 'event field ' + compulsoryFields[i] + ' is missing';

},

list = function(options, cb) {

  if (typeof options == 'function') {

    cb = options;

    options = {};

  }

  var params = lib.extend({
    offset: 0,
    limit: 40,
    order: 'updated_at desc',
    new: false
  }, options);

  var con = model.getConnection();

  // first request gets the ids, second gets the events.

  con.query(['select id, updated_at from event', 'where is_new=false and is_published=true' , 'order by', params.order, 'limit', params.limit, 'offset', params.offset].join(' '), function(err, rows, fields) {

    if (err) return cb(err);

    var ids = rows.map(function(row) { return row.id; });

    if (!ids.length) return cb(null, []);

    var query = [
      getQuery,
      'and e.id in (' + ids.join(',') + ')',
      'order by field(e.id,' + ids.join(',') + ')'
    ];

    con.query(query.join(' '), function(err, rows) {

      model.endConnection();

      if (err) return cb(err);

      var events = [];

      ids.forEach(function(id) {

        events.push(_assembleEvent(id, rows));

      });

      cb(null, events);

    });

  });

},

get = function(params, cb) {

  var con = model.getConnection(),

  criterias = [];

  for (var field in params)
    criterias.push('e.' + lib.toUnderscore(field) + ' = ' + con.escape(params[field]));

  var query = getQuery + ' and ' + criterias.join(' and ');

  con.query(query, function(err, rows, fields) {

    model.endConnection();

    if (err) return cb(err);

    if (!rows.length) return cb(null, null);

    var event = _assembleEvent(rows[0].id, rows);

    cb(null, event);

  });

},

getByLocationId = function(locationId, cb) {

  var con = model.getConnection();

  con.query('select event.id as id from event left join event_location on event.id=event_location.event_id where event_location.location_id = ?', locationId, function(err, results) {

    if (err) throw err;

    cb(results);

  });

},


/**
 * check wether event exists in database by looking at slug, location id, owner id and date
 */

basicCheckExists = function(event, cb) {

  validateSingleDateEvent(event);

  var con = model.getConnection()

  , query = [
    'select e.id from event as e ',
    'left join occurrence as o on e.id=o.event_id ',
    'where slug like ', con.escape(model.slugify(event.title) + '%'), ' ',
    'and o.location_id = ', con.escape(event.locationId), ' ',
    'and o.date = ', con.escape(event.start.getFullYear() + '-' + lib.addZero(event.start.getMonth() + 1) + '-' + lib.addZero(event.start.getDate())), ' ',
    'and e.owner_id = ', con.escape(event.ownerId)
  ].join('');

  con.query(query, function(err, rows, fields) {

    if (err) throw err;

    model.endConnection();

    cb(rows.length?true:false);

  });

},

/**
 * get upcoming events - mega requête
 */

getUpcoming = function(offset, limit, cb) {

  var con = model.getConnection()

  , today = new Date

  , query = [
    'select distinct e.id from event as e ',
    'left join occurrence as o on e.id=o.event_id ',
    'where e.is_new=false and e.is_published=true and o.date >= ', con.escape(today.getFullYear() + '-' + lib.addZero(today.getMonth()+1) + '-' + lib.addZero(today.getDate())), ' ',
    'limit ', offset, ', ', limit
  ].join('');

  con.query(query, function(err, rows, fields) {

    if (err) throw err;

    // foreach event grab the whole set

    var eventIds = rows.map(function(row) { return row.id} ),
      events = [];

    countdown = eventIds.length;

    eventIds.forEach(function(eventId) {

      getEvent(eventId, function(event){

        events.push(event);

        countdown--;

        if (!countdown) {

          model.endConnection();

          cb(events);

        }

      });

    });
  
  });

},


create = function(data, cb) {

  var langs = [],

  descFields = ['title', 'description', 'tags', 'freeText'];

  descFields.forEach(function(field) {

    for (var lang in data[field])
      if (!lib.contains(langs, lang)) langs.push(lang);

  });


  // make inserts in order

  async.waterfall([

    // get unique slug & uids
    function ( wcb ) {

      for (var lang in data.title) break;

      _createSlugAndUid(data.title[lang], wcb);

    },

    // create the event entry
    function ( su, wcb ) {

      model.insert('event', {
        owner_id: data.ownerId,
        slug: su.slug,
        uid: su.uid,
        created_at: new Date(),
        updated_at: new Date(),
        is_published: true,
        is_new: false
      }, function(err, result) {

        if (err) return wcb(err);

        data.slug = su.slug;

        data.uid = su.uid;

        wcb(null, result.insertId);

      });

    },

    // build and create event translation entries
    function ( eventId, wcb ) {

      var tEntries = [];

      langs.forEach(function(lang) {

        var tEntry = {
          id: eventId,
          lang: lang,
        };

        descFields.forEach(function (field) {

          tEntry[lib.toUnderscore(field)] = data[field][lang];

        });

        tEntries.push(tEntry);

      });

      model.insert('event_translation', tEntries, function (err, result) {

        if (err) return wcb(err);

        wcb(null, eventId);

      });

    },

    // fetch location ids by uids for occurrence and event_location inserts
    function ( eventId, wcb ) {

      var lUids = data.locations.map(function(l) { return l.uid; });

      model.getConnection().query('select id, uid from location where uid in  (' + lUids.join(',') + ')', function(err, lRows) {

        if (err) return wcb(err);

        var lIds = {};

        for (var i = lRows.length - 1; i >= 0; i--)
          lIds[lRows[i].uid] = lRows[i].id;

        wcb(null, eventId, lIds);

      });

    },

    // build and create occurrence entries
    function ( eventId, lIds, wcb ) {

      var oEntries = [];

      for (var i = data.locations.length - 1; i >= 0; i--) {
        for (var j = data.locations[i].timings.length - 1; j >= 0; j--) {
        
          var timing = data.locations[i].timings[j];

          oEntries.push({
            event_id: eventId,
            location_id: lIds[data.locations[i].uid],
            date: timing.start.substr(0, 10),
            time_start: timing.start.substr(11),
            time_end: timing.end.substr(11),
            created_at: new Date(),
            updated_at: new Date()
          });

        }
      }

      model.insert('occurrence', oEntries, function (err, result) {

        if (err) return wcb(err);

        wcb(null, eventId, lIds, wcb);

      });

    },

    // build event_location entries
    function ( eventId, lIds, wcb ) {

      var elEntries = [];

      for (var i = data.locations.length -1; i >= 0; i--) {

        elEntries.push({
          event_id: eventId,
          location_id: lIds[data.locations[i].uid],
          ticket_link: data.locations[i].ticketLink,
          created_at: new Date(),
          updated_at: new Date()
        });

      }

      model.insert('event_location', elEntries, function(err, result) {

        if (err) return wcb(err);

        wcb(null, eventId, lIds);

      });

    },

    // map created event_location ids to location ids
    function ( eventId, lIds, wcb ) {

      var elIds = {};

      model.getConnection().query('select id, location_id from event_location where event_id = ' + eventId, function(err, elRows) {

        if (err) return wcb(err);

        elRows.forEach(function(elRow) {
          elIds[elRow.location_id] = elRow.id;
        });

        wcb(null, eventId, lIds, elIds);

      });

    },

    // create event_location_translation entries
    function (eventId, lIds, elIds, wcb) {

      var eltEntries = [];

      data.locations.forEach(function (location) {

        langs.forEach(function (lang) {

          if (location.pricingInfo[lang]) eltEntries.push({
            id: elIds[lIds[location.uid]],
            lang: lang,
            pricing_info: location.pricingInfo[lang]
          });

        });

      });

      model.insert('event_location_translation', eltEntries, wcb);

    }
    
  ], function(err) {

    if (err) return cb(err);

    cb(null, data);

  });

},


/**
 * add a single date / single location event
 *
 * event should have the following fields
 * title, description, start, locationId
 *
 * optionally:
 * freeText, tags, lang (defaults to fr), end
 */

createSingleDateEvent = function(event, cb) {

  validateSingleDateEvent(event);

  // prepare event entry

  var eventEntry = { owner_id: event.ownerId };

  if (event.image) {
    // if the creation is a success, upload to S3. It would also be better to check here that the image actually is where it says it is.
    eventEntry.image = 'event_' + event.image.split('/').pop().split('.')[0] + '_00.jpg';
  }
    
  // prepare event translation entry

  var eventTranslationEntry = {
    lang: event.lang?event.lang:'fr',
    title: event.title,
    description: event.description
  };

  if (event.tags)
    eventTranslationEntry.tags = event.tags;

  if (event.freeText)
    eventTranslationEntry.free_text = event.freeText;


  // prepare event occurrence entry

  var occurrenceEntry = { start: event.start, location_id: event.locationId };

  if (event.end) occurrenceEntry.end = event.end;


  // validate the data
  try {

    validateEventEntry(eventEntry) && validateEventTranslationEntry(eventTranslationEntry) && validateOccurrenceEntry(occurrenceEntry);

  } catch (e) {

    return cb(e);

  }


  var flags = { translation: false, occurrence: false, eventLocation: false, image: false }

  , signalComplete = function() { 

    if (!flags.translation || !flags.occurrence || !flags.eventLocation || !flags.image) return;

    model.endConnection();

    cb(event);

  };

  addEventEntry(event.title, eventEntry, function(createdEventEntry) {

    // event entry is created, can now create occurrence, translation and event_location entries

    eventTranslationEntry.id = createdEventEntry.id;
    occurrenceEntry.event_id = createdEventEntry.id;

    event.id = createdEventEntry.id;

    addEventTranslationEntry(eventTranslationEntry, function(createdEventTranslationEntry) {

      console.log('event - translation entry added');

      flags.translation = true;

      signalComplete();

    });

    addOccurrenceEntry(occurrenceEntry, function(createdOccurrenceEntry) {

      console.log('event - occurrence entry added');

      flags.occurrence = true;

      signalComplete();

    });

    addEventLocationEntry({ event_id: createdEventEntry.id, location_id: event.locationId}, function(createdEventLocationEntry) {

      console.log('event - eventLocation entry added');

      addEventLocationTranslationEntry({
        event_location_id: createdEventLocationEntry.id
      }, function() {

        console.log('event - eventLocationTranslation entry added');

        flags.eventLocation = true;

        signalComplete();
      });

    });

    if (event.image) {

      console.log('event has image: ' + event.title + ' image: ' + event.image);

      var imagePath = event.image.split('/')
        , imageName = imagePath.pop()
        , thumbnailName = thumbnailPrefix + imageName;
      
      imagePath = imagePath.join('/');

      // resize pic to 300px width
      image.resize(event.image, event.image, { width: 300 }, function(success){

        if (!success) {
          flags.image = true;
          return signalComplete();
        }

        // create thumbnail
        image.thumbnailize(event.image, imagePath + '/' + thumbnailName, {width: 120, height: 160}, function() {

          // store images
          fileHandler.store(imagePath, [{src: imageName, dst: eventEntry.image}, {src: thumbnailName, dst: thumbnailPrefix + eventEntry.image}], function(err) {
            flags.image = true;
            signalComplete();
          });

        });

      });

    } else {
      flags.image = true;
      signalComplete();
    }

    // create image thumbnail, then upload both to S3

  });

},


/**
 * generates unique event slug and uid
 */

_createSlugAndUid = function(title, cb) {

  var slug = false;

  // first get a slug
  model.getFieldNewUniqueValue('event', 'slug', function () {
    
    slug = model.slugify(title) + (slug?Math.ceil(Math.random()*10000):'');

    return slug;

  }, function ( slug ) {

    // get a uid
    model.getFieldNewUniqueValue('event', 'uid', function () {

      return Math.ceil(Math.random()*99999999);

    }, function ( uid ) {

      cb(null, {slug: slug, uid: uid});

    });

  });

},

/**
 * add row information to event object
 */

_assembleEvent = function( eventId, rows ) {

  var event, 

  processed = { languages: [], locations: [], eventLocations: {} };

  rows.filter(function(row) { return row.id==eventId; }).forEach(function(row) {

    if (!event) {

      event = {
        id: row.id,
        ownerId: row.owner_id,
        slug: row.slug,
        createdAt: row.created_at,
        updatedAt: row.updated_at,
        uid: row.uid,
        title: {},
        description: {},
        tags: {},
        freeText: {},
        locations: []
      };

      if (row.image) {

        event.image = s3Path + '/' + row.image;
        event.thumbnail = s3Path + '/' + thumbnailPrefix + row.image;
      
      }

    }

    // extract language fields
    if (!lib.contains(processed.languages, row.lang)) {

      processed.languages.push(row.lang);

      event.title[row.lang] = row.title;
      event.description[row.lang] = row.description;
      if (row.free_text) event.freeText[row.lang] = row.free_text;
      if (row.tags) event.tags[row.lang] = row.tags;

    }

    // extract location

    var locationIndex;

    if (!lib.contains(processed.locations, row.location_id)) {

      var newLocation = {
        uid: row.location_uid,
        verified: row.location_verified,
        name: row.placename,
        address: row.address,
        city:row.city,
        department:row.department,
        region:row.region,
        district:row.district,
        postcode : row.postcode,
        latitude: row.latitude,
        longitude: row.longitude,
        timings: [],
      };

      if (row.ticket_link) newLocation.ticketLink = row.ticket_link;

      event.locations.push(newLocation);

      processed.eventLocations[row.location_id] = row.ellang;

      processed.locations.push(row.location_id);

      locationIndex = processed.locations.length - 1;

    }

    if (typeof locationIndex == 'undefined')
      for (var index in processed.locations)
        if (processed.locations[index] == row.location_id) locationIndex = index;

    // get pricing info

    if (row.pricing_info!==null && row.pricing_info.length) {
      if (!event.locations[locationIndex].pricingInfo) event.locations[locationIndex].pricingInfo = {};
      if (!event.locations[locationIndex].pricingInfo[row.ellang]) event.locations[locationIndex].pricingInfo[row.ellang] = row.pricing_info;
    }

    if (row.date && (processed.eventLocations[row.location_id] == row.ellang)) {

      // extract timing at location at date

      var start = new Date(row.date.getFullYear() + '-' + lib.addZero(row.date.getMonth()+1) + '-' + lib.addZero(row.date.getDate()) + 'T' + row.time_start.substr(0,5)),
      
      end = row.time_end?new Date(row.date.getFullYear() + '-' + lib.addZero(row.date.getMonth()+1) + '-' + lib.addZero(row.date.getDate()) + 'T' + row.time_end.substr(0,5)):new Date(row.date.getFullYear() + '-' + lib.addZero(row.date.getMonth()+1) + '-' + lib.addZero(row.date.getDate()+1));

      if (end<start) end = new Date(end.getTime() + 24*60*60*1000);

      event.locations[locationIndex].timings.push({ start: start, end: end });
    }
      
  });

  if (event.freeText && !lib.size(event.freeText)) delete event.freeText;
  if (event.tags && !lib.size(event.tags)) delete event.tags;

  return event;

};
