// cibulModel/lib/locations.js
module.exports = function(m, f) {

  model = m;

  fileHandler = f;

  return {
    create: create,
    createOrGet: createOrGet,
    clear: clear,
    list: list,
    get: get,
    getNeighbors: getNeighbors,
    total: total,
    update: update,
    merge: merge
  };
  
};


var model, 

lib = require('./lib'),

php = require('./php'), 

image = require('./image'), 

fileHandler,


/**
 * clear the location set
 */

clear = function(cb) {

  model.clear([
    'location',
    'event_location',
    'occurrence'
  ], cb);

},


/**
 * get one location
 */

get = function(params, cb) {

  var con = model.getConnection(),
  query = 'select * from location where ',
  where = [],
  fields = {};

  for (var field in params) {
    // params[field] contains the value(s)
    if (typeof params[field] !== 'object') {

      fields[field] = [lib.contains(['latitude', 'longitude', 'uid', 'id', 'main'], field)?params[field]:con.escape(params[field])];

    } else {

      fields[field] = [];

      for (var i = params[field].length - 1; i >= 0; i--) {

        var value = params[field][i];

        fields[field].push(lib.contains(['latitude', 'longitude', 'uid', 'id', 'main'], field)?value:con.escape(value));

      }

    }
  }

  for (field in fields)
    where.push(field + ' in (' + fields[field].join(',') + ')');

  query = query + where.join(' and ');

  con.query(query, function(err, rows, fields) {

    model.endConnection();

    if (err) return cb(err);

    if (rows.length) rows.forEach(function(row) {
      row.store = row.store===null?{}:php.unserialize(row.store);
    });

    cb(null, rows.length?rows[0]:null);

  });

},


/**
 * create a location
 */

create = function(values, cb) {

  _validateLocationEntry(values);

  var con = model.getConnection(),

  entry = {
    placename: values.name,
    latitude: values.latitude,
    longitude: values.longitude,
    owner_id: values.ownerId,
    updated_at: new Date(),
    created_at: new Date()
  },

  slug = false;

  ['address', 'city', 'country'].forEach(function(field) { if (values[field]) entry[field] = values[field]; });

  if (values.main) entry.main = !!values.main;

  if (values.store) entry.store = php.serialize(values.store);

  model.getFieldNewUniqueValue('location', 'uid', function() {
    return Math.ceil(Math.random()*99999999);
  }, function(uid) {

    model.getFieldNewUniqueValue('location', 'slug', function() {

      slug = model.slugify(values.name) + (slug?Math.ceil(Math.random()*10000):'');

      return slug;

    }, function(slug) {

      entry.uid = uid;
      entry.slug = slug;

      model.getConnection().query('insert into location set ?', entry, function(err, result) {

        if (err) return cb(err);

        entry.id = result.insertId;

        model.endConnection();

        cb(null, entry);

      });


    });

  });

},

_validateLocationEntry = function(values) {

  if (!values.name) throw "Location must have a name";
  if (!values.ownerId) throw "Location must have an owner_id";
  if (!values.latitude) throw "Location must have a latitude";
  if (!values.longitude) throw "Location must have a longitude";
  if (!values.address) throw "Location must have an address";

},

_buildListQuery = function(con, fields, params) {

  if (typeof fields == 'string') fields = [fields];

  var query = ['select', fields.join(', '), 'from location'];

  // add filters

  var filters = [];

  // main
  if (params.main !== undefined && params.main.length)
    filters.push('main = ' + con.escape(params.main=='true'?true:false));

  // placename
  if (params.placename !== undefined && params.placename.length)
    filters.push('placename like ' + con.escape('%' + params.placename + '%'));

  // address
  if (params.address !== undefined && params.address.length)
    filters.push('address like ' + con.escape('%' + params.address + '%'));

  // processed at
  if (params.unprocessed)
    filters.push('processed_at is null');

  return filters.length?query.concat(['where'], [filters.join(' and ')]):query;

},

total = function(params, cb) {

  var con = model.getConnection();

  con.query(_buildListQuery(con, 'count(id) as total_count', params).join(' '), function(err, rows, fields) {

    if (err) throw err;

    cb(rows[0].total_count);

  });

},

getNeighbors = function(latitude, longitude, params, cb) {

  params = lib.extend({
    accurracy: 0.001,
    exclude: []
  },params);

  var con = model.getConnection(),

  /**
   decimal places   degrees      distance
       0         1            111   km
       1         0.1          11.1  km
       2         0.01         1.11  km
       3         0.001        111   m
       4         0.0001       11.1  m
       5         0.00001      1.11  m
       6         0.000001     0.111 m
       7         0.0000001    1.11  cm
       8         0.00000001   1.11  mm

       params.location (latitude, longitude,)
      params.accurracy (number between 3 and 5)
  */

  query = [
    'select id, placename, address, latitude, longitude, uid, main,',
    '( 6371 * acos( sin( latitude / 57.2958 ) * sin(', latitude, '/ 57.2958 )',
    '+ cos( latitude / 57.2958 ) * COS(', latitude, '/ 57.2958 )',
    '* cos(', longitude, '/ 57.2958 - longitude / 57.2958 ))) *1000 AS distance',
    'from location where location.latitude >', latitude - params.accurracy,
    'and location.latitude <', latitude + params.accurracy,
    'and location.longitude >', longitude - params.accurracy,
    'and location.longitude <', longitude + params.accurracy
  ];

  if (params.exclude.length) query.push('and location.id not in (' + params.exclude.join(',') + ')');

  query.push('order by distance asc');

  con.query(query.join(' '), function(err, rows, fields) {

    if (err) throw err;

    cb(rows);

  });

},

list = function(params, cb) {

  params = lib.extend({
    page: 1,
    limit: 40,
    order: ''
  }, params);

  var con = model.getConnection(),

  query = _buildListQuery(con, '*', params);

  if (params.order.length) query.push('order by', params.order);

  con.query(query.concat(['limit', params.limit*(params.page-1), ',', params.limit]).join(' '), function(err, rows, fields) {

    if (err) throw err;

    cb(rows);

  });

},

_processImage = function(location, path, cb) {

  path = path.split('/'), tmpName = path.pop(), name = 'location_' + location.uid + Math.ceil((Math.random()*1000)) + '.jpg',

  path = path.join('/');

  image.resize(path + '/' + tmpName, path + '/' + tmpName, { width: 300 }, function(success){
    
    // store images
    fileHandler.store(path, [{src: tmpName, dst: name}], function(err) {
      
      // if all went well and there was an image before, it should be removed
      if (!err && location.store && location.store.image) {
        fileHandler.remove(location.store.image, function(err) { cb(err, name); });
      } else {
        cb(err, name);
      }

    });

  });

},

update = function(item, values, cb) {

  var _runUpdate = function(err, image) {

    if (err) throw err;

    var con = model.getConnection(),

    query = [];

    if (values.store) {

      values.store = lib.extend(item.store, values.store);

      if (image)
        values.store.image = image;
      
      values.store = php.serialize(values.store);
    }

    values.updated_at = new Date();

    for (var fieldName in values)
      query.push(fieldName + '=' + con.escape(values[fieldName]));

    query = 'update location set ' + query.join(', ') + ' where id=' + con.escape(item.id);

    con.query(query, function(err, result) {

      if (err) throw err;

      cb(true);

      model.endConnection();

    });

  };

  // if there is an image, need to name it, upload it to s3, remove previous image and set new value if all that went ok;

  if (values.store && values.store.image && values.store.image.length) {
    _processImage(item, values.store.image, _runUpdate);
  }
  else
    _runUpdate(null);

},

/**
 * replace all references to location (params.id) with references to main location (params.mainId) and remove location.
 */

merge = function(params, cb) {

  var con = model.getConnection(),

  processEventLocation = function(eventLocations, iteration, onComplete) {

    if (!eventLocations.length) return onComplete();

    var eventLocation = eventLocations[iteration],

    next = function(err, results) {

      if (++iteration == eventLocations.length)
        onComplete();
      else
        processEventLocation(eventLocations, iteration, onComplete);

    };

    // if there is an event location linking event to main location, we remove this
    // else, we update it

    con.query('select id from event_location where event_id=? and location_id=?', [eventLocation.event_id, params.mainId], function(err, rows){

      if (rows.length) {

        con.query('delete from event_location where id=?', eventLocation.id, next);

      } else {

        con.query('update event_location set location_id=? where id=?', [params.mainId, eventLocation.id], next);

      }

    });

  };

  get({id: [params.mainId, params.id]}, function(results) {

    if (results.length!=2) throw "Could not retrieve both locations";

    con.query('select id, event_id from event_location where location_id=' + params.id, function(err, rows) {

      if (err) throw err;

      processEventLocation(rows, 0, function() { // go through all event locations and update references
        
        // go through all occurrences and update references
        con.query('update occurrence set location_id=? where location_id=?', [params.mainId, params.id], function(err, result) {

          if (err) throw err;

          // remove old location

          con.query('delete from location where id=?', params.id, function(err, result) {

            if (err) throw err;

            cb(true);

            model.endConnection();

          });

        });

      });

    });

  });

},

_matchByNameAndLocation = function(name, latitude, longitude, cb) {

  var con = model.getConnection(),

  query = [
    'select * from location where latitude like ',
    con.escape(('' + latitude).substr(0,6) + '%'),
    ' and longitude like ',
    con.escape(('' + longitude).substr(0,6) + '%'),
    ' and placename = ',
    con.escape(name)
  ].join('');

  con.query(query, function(err, rows, fields) {

    if (err) throw err;

    cb(rows.length?rows[0]:false);

  });

},



createOrGet = function(values, cb) {

  try {
    _validateLocationEntry(values);
  }
  catch (e) {
    return cb(e);
  }

  _matchByNameAndLocation(values.name, values.latitude, values.longitude, function(result) {

    if (!result) {
      create(values, function(result) {
        model.endConnection();
        cb(result);
      });
    } else {
      model.endConnection();
      cb(result);
    }
      

  });

};
