// cibulModel/lib/model.js
module.exports = function( config ) {

  dbParams = config;

  return {
    getConnection: getConnection,
    endConnection: endConnection,
    slugify: slugify,
    getFieldNewUniqueValue: getFieldNewUniqueValue,
    remove: remove,
    clear: clear,
    insert: insert
  };

};

var mysql = require('mysql'),

async = require('async'),

con,

util = require('util'),

lib = require('./lib'),

getConnection = function() {

  if (!con) con = mysql.createConnection(dbParams);

  return con;

},

endConnection = function() {

  if (con) con.end();

  con = false;

},

slugify = function( text ) {
  return text
    .toLowerCase()
    .replace(/ /g,'-')
    .replace(/[^\w-]+/g,'');
},


/**
 * tests table field to value given by generator function, calls back value when not already existing
 */

getFieldNewUniqueValue = function( table, field, generator, cb ) {

  var value = generator(),
      con = getConnection(),
      query = ['select ', field, ' from ', table, ' where ', field, ' = ', con.escape(value)].join('');

  con.query(query, function(err, rows, fields) {

    if (err) throw err;

    if (rows.length)
      getFieldNewUniqueValue(table, field, generator, cb);
    else
      cb(value);
  });

},


/**
 * clear a table or a bunch of tables
 */

clear = function clear( table, cb, close ) {

  if (typeof close == 'undefined') close = true;

  if (typeof table !== 'string') {

    async.parallel(table.map(function(name) {

      return function(pCb) {

        clear(name, pCb, false);

      };

    }), function(err) {

      endConnection();

      if (err) return cb(err);

      cb();

    });

  } else {

    getConnection().query('delete from ' + table, function(err) {

      if (close) endConnection();

      if (err) return cb(err);

      cb();

    });

  }

},


/**
 * delete entries matching whereFields values
 */

remove = function( table, whereFields, cb ) {

  var where = [];

  for (field in whereFields) 
    where.push(field + '=' + getConnection().escape(whereFields[field]));

  getConnection().query(['delete from ', table, ' where ', where.join(' and ')].join(''), function(err, result) {
    if (err) throw err;

    if (cb) cb();
  });

},

insert = function( table, data, cb ) {

  if (!lib.isArray(data)) data = [data];

  var keys = [];

  for (var k in data[0])
    keys.push(k);

  var query = 'insert into ' + table + ' (' + keys.join(',') + ') values ';

  for (var i = 0; i < data.length; i++) {

    var values = [];

    for (var j = 0; j < keys.length; j++) {

      values.push(getConnection().escape(data[i][keys[j]]));

    }

    query += '(' + values.join(', ') + '),';
    
  };

  query = query.slice(0, - 1) + ';'

  getConnection().query(query, function(err, result) {

    endConnection();

    cb(err, result);

  });

};
