// cibulModel/lib/reviews.js
module.exports = function(m) {

  model = m;

  return {
    list: list,
    get: get,
    create: create,
    clear: clear,
    total: total,
    getFields: getFields,
    addEvent: addEvent
  };
  
};

var model, 

getQuery = [
  'select * from review as r'
],

lib = require('./lib'),

async = require('async'),

get = function(params, cb) {

  var con = model.getConnection(),

  criterias = [];

  for (var field in params)
    criterias.push('r.' + lib.toUnderscore(field) + ' = ' + con.escape(params[field]));

  var query = getQuery + ' where ' + criterias.join(' and ');

  con.query(query, function(err, rows, fields) {

    model.endConnection();

    if (typeof cb == 'undefined') return;

    if (err) return cb(err);

    if (!rows.length) return cb(null, null);

    cb(null, rows[0]);

  });

},

create = function(data, cb) {

  async.waterfall([

    // get unique uid
    
    function ( wcb ) {

      model.getFieldNewUniqueValue('review', 'uid', function() {

        return Math.ceil(Math.random()*99999999);

      }, function ( uid ) {

        wcb(null, uid);

      });

    },

    // get unique slug

    function ( uid, wcb ) {

      var slug;

      model.getFieldNewUniqueValue('review', 'slug', function() {

        slug = model.slugify(data.title) + (slug?Math.ceil(Math.random()*10000):'');

        return slug;

      }, function ( slug ) {

        wcb(null, uid, slug);

      });

    },


    // create the thing
    
    function ( uid, slug, wcb ) {

      model.insert('review', {
        owner_id: data.ownerId,
        slug: slug,
        uid: uid,
        created_at: new Date(),
        updated_at: new Date(),
        title: data.title,
        description: data.description,
        url: data.url
      }, function(err, result) {

        if (err) return cb(err);

        data.id = result.insertId;

        data.slug = slug;

        data.uid = uid;

        cb(null, result.insertId);

      });

    }

  ]);

},


/**
 * clear the review set
 */

clear = function(cb) {

  model.clear([
    'review_article',
    'review_category',
    'review_credential',
    'review_embed',
    'review_embed_facebook_page',
    'review_tag',
    'review_tag_article',
    'reviewer',
    'aggregator',
    'aggregator_source',
    'review'
  ], cb);

},

_buildListQuery = function(con, fields, params) {

  if (typeof fields == 'string') fields = [fields];

  var queryFields = [];

  ['id', 'uid', 'title', 'description', 'owner_id', 'email', 'created_at'].forEach(function(fieldName) {

    if (!lib.contains(fields, fieldName)) return;

    if (lib.contains(['id', 'created_at'], fieldName)) fieldName = 'review.' + fieldName + ' as ' + fieldName;

    queryFields.push(fieldName);

  });

  if (lib.contains(fields, 'total_count')) queryFields.push('count(review.id) as total_count');

  var query = ['select', queryFields.join(', '), 'from review'];

  query.push('left join user on review.owner_id=user.id');

  var filters = [];

  if (params.ownerId !== undefined)
    filters.push('owner_id = ' + con.escape(params.ownerId));

  if (params.email !== undefined && params.email.length)
    filters.push('email like ' + con.escape('%' + params.email + '%'));

  if (params.title !== undefined && params.title.length)
    filters.push('title like ' + con.escape('%' + params.title + '%'));

  if (filters.length) query = query.concat(['where'], [filters.join(' and ')]);

  query.push('order by review.created_at desc');

  return query;

},

addEvent = function(reviewId, eventId, userId, callback, params) {

  var con = model.getConnection(),

  // only exit.
  conclude = function(result) {

    model.endConnection();

    callback(result);

  };

  con.query('select id from review where id=' + con.escape(reviewId), function(err, rows, fields) {

    if (!rows.length) {

      return conclude(false);

    }

    con.query('select id from event where id=' + con.escape(eventId), function(err, rows, fields) {

      if (!rows.length) {

        return conclude(false);

      }

      con.query('select id from review_article where review_id=' + con.escape(reviewId) + ' and event_id=' + con.escape(eventId), function(err, rows, fields) {

        if (rows.length) {

          return conclude(false);
          
        }

        con.query('select id from user where id=' + con.escape(userId), function(err, rows, fields) {

          if (!rows.length) {

            return conclude(false);

          }

          var entry = {
            user_id: userId,
            review_id: reviewId,
            event_id: eventId,
            created_at: new Date(),
            updated_at: new Date()
          };

          if (params.category) entry.category = params.category;

          con.query('insert into review_article set ?', entry, function(err, result) {
            if (err) throw err;

            conclude(true);
          });

        });

      });

    });

  });

},

total = function(params, callback) {

  var con = model.getConnection();

  con.query(_buildListQuery(con, 'count(id) as total_count', params).join(' '), function(err, rows, fields) {

    if (err) throw err;

    callback(rows[0].total_count);

  });

},

list = function(params, callback) {

  params = lib.extend({
    page: 1,
    limit: 40,
    fields: '*'
  }, params);

  var con = model.getConnection(),

  query = _buildListQuery(con, params.fields, params);

  if (params.limit) query = query.concat(['limit', params.limit*(params.page-1), ',', params.limit]);

  con.query(query.join(' '), function(err, rows, fields) {

    if (err) throw err;

    callback(rows);

  });

},

getFields = function() {
  return {
    uid: 'integer',
    title: 'string',
    description: 'title',
  };
};
