// cibulModel/lib/reviewCredentials.js
module.exports = function(config) {

  model = require('./model')(config);

  return {
    get: get,
    total: total,
    list: list,
    create: create,
    remove: remove,
    update: update
  };
  
};

var model, lib = require('./lib'),

_buildListQuery = function(con, fields, params) {

  var cleanedFields = [], joins = [];

  if (typeof fields == 'string') {

    cleanedFields = [fields];

  } else {

    lib.forEach(fields, function(field) {

      if (lib.contains(['id', 'type', 'owner_id'], field)) {

        cleanedFields.push('rc.' + field + ' as ' + field);

      } else if (field=='ownerEmail') {

        cleanedFields.push('u.email as ownerEmail');

        joins.push('left join user as u on rc.owner_id = u.id');

      } else if (field=='reviewTitle') {

        cleanedFields.push('r.title as reviewTitle');

        joins.push('left join review as r on rc.review_id = r.id');

      }

    });

  }

  var query = ['select', cleanedFields.join(', '), 'from review_credential as rc'].concat(joins);

  // add filters

  var filters = [];
  
  if (params.ownerId !== undefined)
    filters.push('owner_id = ' + con.escape(params.ownerId));

  if (filters.length) query = query.concat(['where'], [filters.join(' and ')]);

  console.log(query.join(' '));

  return query;

},



total = function(params, callback) {

  var con = model.getConnection();

  con.query(_buildListQuery(con, 'count(id) as total_count', params).join(' '), function(err, rows, fields) {

    if (err) throw err;

    callback(rows[0].total_count);

  });

},


list = function(params, callback) {

  var con = model.getConnection(),

  query = _buildListQuery(con, params.fields?params.fields:'*', params).join(' ');

  con.query(query, function(err, rows, fields) {

    if (err) throw err;

    callback(rows);

  });

},

get = function(params, callback) {

  var con = model.getConnection();

  if (!params.id) return callback('this is only possible via id');

  con.query('select * from review_credential where id=' + params.id, function(err, rows, fields) {

    if (err) return callback(err);

    callback(null, rows);

  });

},

update = function(item, values, callback) {

  var con = model.getConnection(), query = [];

  if (values.updated_at === null) values.updated_at = new Date();

  for (var fieldName in values)
    query.push(fieldName + '=' + con.escape(values[fieldName]));

  con.query('update review_credential set ' + query.join(', ') + ' where id=' + con.escape(item.id), function(err, result) {

    if (err) return callback(err);

    callback(null, result);

    model.endConnection();

  });

},

create = function(values, callback) {

  if (!values.ownerId) return callback('ownerId missing');

  var newEntry = { owner_id: values.ownerId };

  var type = null;

  if (values.type=='community') {
    
    lib.extend(newEntry, {
      type: 1,
      integrated: true,
      newsletter: true,
      facebook: true,
      admins: true,
      embedded_form: true,
      moderated: true,
      extended_rights: true,
      aggregator: true,
      indesign: true,
      dataviz: true,
      editors: true
    });

  } else if (values.type=='editor') {

    lib.extend(newEntry, {
      type: 0,
      integrated: true,
      newsletter: true,
      facebook: true,
      admins: true
    });

  }

  model.getConnection().query('insert into review_credential set ?', newEntry, function(err, result) {

    if (err) return callback(err);

    newEntry.id = result.insertId;

    model.endConnection();

    callback(null, newEntry);

  });

},

remove = function(item, callback) {

  var con = model.getConnection();

  if (item.id === undefined) return callback('id not specified');

  con.query('delete from review_credential where id=' + con.escape(item.id), function(err, result) {

    if (err) return callback(err);

    callback(null);

    model.endConnection();

  });

};
