// cibulModel/lib/tokens.js
module.exports = function(config) {

  model = require('./model')(config);

  return {
    getUserActivationToken: getUserActivationToken
  };

}

var model, lib = require('./lib'),

TYPES = {
  ACCOUNTACTIVATE: 'aa',
},

getUserActivationToken = function(userId, email, forceCreate, callback) {

  if (typeof forceCreate == 'function') {
    callback = forceCreate;
    forceCreate = false;
  }

  _getByTypeAndEmail(TYPES.ACCOUNTACTIVATE, email, function(token) {

    if (!forceCreate || token) return callback(token.token);

    var con = model.getConnection(),

    entry = {
      user_id: userId,
      email: email,
      type: TYPES.ACCOUNTACTIVATE,
      token: require('crypto').createHash('sha1').update(email).digest("hex")
    };

    con.query('insert into user_token set ?', entry, function(err, result) {

      if (err) throw err;

      callback(entry.token);

    });

  });

},

_getByTypeAndEmail = function(type, email, callback) {

  var con = model.getConnection();

  con.query('select * from user_token where type=? and email=?', [TYPES.ACCOUNTACTIVATE, email], function(err, results) {
    
    if (err) throw err;

    callback(results.length?results[0]:false);

  });

};
