// cibulModel/lib/users.js
module.exports = function(m, f) {
  
  model = m;

  return {
    validateEmailAndPassword: validateEmailAndPassword,
    getFields: getFields,
    total: total,
    list: list,
    get: get,
    create: create,
    update: update
  };
  
};

var model,

lib = require('./lib'),

total = function(params, cb) {

  model.getConnection().query(_buildListQuery('count(u.id) as total_count', params).join(' '), function(err, rows, fields) {

    model.endConnection();

    if (err) return cb(err);

    cb(null, rows[0].total_count);

  });

},

list = function(params, cb) {

  params = lib.extend({
    page: 1,
    limit: 40
  }, params);

  var query = _buildListQuery('u.*, count(rc.type=0) as editor, count(rc.type=1) as community', params).concat(['group by u.id', 'order by u.created_at desc', 'limit', params.limit*(params.page-1), ',', params.limit]).join(' ');

  model.getConnection().query(query, function(err, rows, fields) {

    model.endConnection();

    if (err) return cb(err);

    cb(null, rows);

  });

},

get = function(params, cb) {

  var lookupField,

  query,

  values;

  if (params.id) {
    
    lookupField = 'u.id';

    params[lookupField] = params.id;

  } else if (params.email) {

    lookupField = 'email';

  } else {

    return cb('Cannot retrieved using given field');

  }

  values = (typeof params[lookupField] !== 'object'?[params[lookupField]]:params[lookupField]).map(function(value) { return model.getConnection().escape(value); });

  query = 'select u.*, count(rc.type=0) as editor, count(rc.type=1) as community from user as u left join review_credential as rc on u.id=rc.owner_id where ' + lookupField + ' in (' + values.join(',') + ') group by u.id;';

  model.getConnection().query(query, function(err, rows, fields) {

    model.endConnection();

    if (err) return cb(err);

    if (!rows.length) return cb(null, null);

    cb(null, rows);

  });

},

update = function(item, values, cb) {

  var query = [],

  fieldName;

  values.updated_at = new Date();

  for (fieldName in values)
    query.push(fieldName + '=' + model.getConnection().escape(values[fieldName]));

  query = 'update user set ' + query.join(', ') + ' where id=' + model.getConnection().escape(item.id);

  model.getConnection().query(query, function(err, result) {

    model.endConnection();

    if (err) return cb(err);

    cb();

  });

},

/**
 * creates a new user with its main review.
 */

create = function(values, cb) {

  var salt = _makeSalt(values.password),

  entry = {
    full_name: values.fullName,
    email: values.email,
    culture: values.culture?values.culture:'en',
    is_activated: !!values.activated,
    salt: salt,
    password: _makeHash(values.password, salt),
    created_at: new Date(),
    updated_at: new Date()
  };

  _createUsername(values.fullName, function(username) {

    entry.username = username;

    model.getConnection().query('insert into user set ?', entry, function(err, result) {

      if (err) {

        model.endConnection();

        return cb(err);

      }

      var userId = result.insertId;

      _createReviewSlugAndUid(entry.full_name, function(reviewSlug, reviewUid) {

        var reviewEntry = {
          title: entry.full_name,
          slug: reviewSlug,
          uid: reviewUid,
          main: true,
          owner_id: userId,
          created_at: new Date(),
          updated_at: new Date()
        };

        model.getConnection().query('insert into review set ?', reviewEntry, function(err, reviewResult) {

          if (err) {

            model.remove('user', {id: userId});

            model.endConnection();

            return cb(err);

          }

          cb(null, userId);

        });

      });

    });

  });

},


_buildListQuery = function(fields, params) {

  if (typeof fields == 'string') fields = [fields];

  var query = ['select', fields.join(', '), 'from user as u left join review_credential as rc on u.id=rc.owner_id'];

  // add filters

  var filters = [];
  
  if (params.email !== undefined && params.email.length)
    filters.push('email like ' + model.getConnection().escape('%' + params.email + '%'));

  if (params.activated !== undefined && params.activated.length)
    filters.push(params.activated=='true'?'is_activated = true':'is_activated = false');

  if (filters.length) query = query.concat(['where'], [filters.join(' and ')]);

  return query;

},


/**
 * generate a unique uid for review
 */

_createReviewSlugAndUid = function(title, cb) {

  var slug = false;

  // first get a slug
  model.getFieldNewUniqueValue('review', 'slug', function() {
    
    slug = model.slugify(title) + (slug?Math.ceil(Math.random()*10000):'');

    return slug;

  }, function(slug) {

    // get a uid
    model.getFieldNewUniqueValue('review', 'uid', function() {

      return Math.ceil(Math.random()*99999999);

    }, function(uid) {

      cb(slug, uid);

    });

  });

},


/**
 * generate a unique username
 */

_createUsername = function(fullName, cb) {

  var username;

  model.getFieldNewUniqueValue('user', 'username', function() {
    
    username = model.slugify(fullName) + (username?Math.ceil(Math.random()*10000):'');

    return username;

  }, cb);

},


/**
 * retrieve a user based on his email
 */

_findOneByEmail = function(email, cb) {

  var query = ['select * from user where is_activated=true and email=', model.getConnection().escape(email)].join('');

  model.getConnection().query(query, function(err, rows, fields) {

    model.endConnection();

    if (err) return cb(err);

    if (!rows.length) return cb(null, null);

    cb(null, rows[0]);

  });

},

_makeSalt = function(password) {

  var chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz',
  
  salt = '';
  
  for (var i=0; i<32; i++) {
    var rnum = Math.floor(Math.random() * chars.length);
    salt += chars.substring(rnum,rnum+1);
  }

  return salt;
  
},

_makeHash = function(password, salt) {

  return require('crypto').createHash('md5').update(password + salt).digest("hex");

},



validateEmailAndPassword = function(email, password, cb) {

  _findOneByEmail(email, function(err, user) {

    model.endConnection();

    if (err) return cb(err);

    if (!user) return cb(null, false);

    return cb(null, user.password==require('crypto').createHash('sha1').update(user.salt+password, 'utf-8').digest('hex')?user:false);

  });
  

},

getFields = function() {
  return {
    is_activated: 'boolean',
    username: 'string',
    full_name: 'string',
    email: 'string',
    community: 'integer',
    editor: 'integer'
  };
};
