// labels/home/index.js
module.exports = {
  myAgendas: {
    fr: 'Mes agendas',
    en: 'My agendas'
  },
  createAgenda: {
    fr: 'Créer un agenda',
    en: 'Create an agenda'
  },
  searchAgenda: {
    fr: 'Rechercher un agenda',
    en: 'Search an agenda'
  },
  myEvents: {
    fr: 'Mes événements',
    en: 'My events'
  },
  messages: {
    fr: 'Messages',
    en: 'Messages'
  },
  notifications: {
    fr: 'Notifications',
    en: 'Notifications'
  },
  addAnEvent: {
    fr: 'Ajouter un événement',
    en: 'Add an event'
  },
  manage: {
    fr: 'Gérer',
    en: 'Manage'
  },
  moderate: {
    fr: 'Modérer',
    en: 'Moderate'
  },
  officialAgenda: {
    fr: 'Agenda officiel',
    en: 'Official agenda'
  },
  welcome: {
    fr: 'Bienvenue sur OpenAgenda !',
    en: 'Welcome to OpenAgenda !'
  },
  orContributeToExisting: {
    fr: 'Ou contribuez à un agenda existant',
    en: 'Or contribute to an existing agenda'
  },
  noResult: {
    fr: 'Aucun résultat trouvé',
    en: 'No result found'
  },
  myEvents: {
    fr: 'Mes événements',
    en: 'My events'
  },
  createEvent: {
    fr: 'Créer un événement',
    en: 'Create an event'
  },
  searchEvent: {
    fr: 'Rechercher un événement',
    en: 'Search an event'
  },
  selectAgenda: {
    fr: 'Séléctionner l\'agenda dans lequel créer l\'événement',
    en: 'Select an agenda in which to create the event'
  },
  privateAgenda: {
    fr: 'Agenda privé',
    en: 'Private agenda'
  },
  privateEvent: {
    fr: 'Événement privé',
    en: 'Private event'
  },
  draft: {
    fr: 'Brouillon',
    en: 'Draft'
  }
};
